Add breadcrumbs for editors

colin-grant-work opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

<!--
Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

This PR updates #6731 and completes the remaining to-do's from that PR.

From that description:
> This commit adds a breadcrumbs bar to the editor widget. It shows the path to the current file and outline information as breadcrumbs. A click of breadcrumbs allows to switch to siblings.

> This fixes #5475.

> How to test:
> Open an editor shows the breadcrumbs on top of the editor. Folders with files are clickable. A popup allows to select a file that will be opened immediately. Outline information is added to the breadcrumbs as well.

![image](https://user-images.githubusercontent.com/62660806/129743733-50a18d68-dd0d-4c5a-b901-664f84bbee65.png)


A few more details:
1. You should see breadcrumbs for `NavigatableWidget`s (mainly editors) that are docked in an area with a horizontal tabbar. Those breadcrumbs should contain at least the path to the file.
2. For editors with language support, you should get the content of the outline view for the location of your cursor.
  > Try refreshing when you have a TS file open. The layout should be restored. There were changes to the Outline view that required changes to its state restoration.
3. If you open a non-navigatable widget, the breadcrumbs bar should disappear.
4. There is a preference to disable breadcrumbs. If you disable the breadcrumbs, they should disappear, and they shouldn't reappear anywhere.
  > Watch out for any odd styling behavior in the transition from breadcrumbs <-> no breadcrumbs. I think I've handled it, but there may still be some quirks.
5. If you click a breadcrumb, it should open a popup. The popup should show other possibilities at the same level of organization (sibling folders, object fields, etc.).
6. You should be able to navigate the tree

#### Review checklist

- [ ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

kenneth-marut-work reviewed (commented):
Looks really nice so far, and really nice functionality. Things mostly seem to be behaving as expecting but I have a few observations after doing an initial feature test:

---
-  Toggling the preference to show hide doesn't seem to resize the base editor, so you get a empty bar: 

![breadrumbs-showhide](https://user-images.githubusercontent.com/62660714/129807732-feea0355-ffe4-448f-915c-5fd014a5a3f8.gif)
---
After waiting several minutes, I was able to get the correct behavior, but it seems to be very finicky. For example in in this GIF, you'll see the breadcrumbs toggle ON, but when I click on the editor they disappear again:

![breadcrumbsnotshowingup](https://user-images.githubusercontent.com/62660714/129809429-d161abed-8fd9-4e64-b54f-debb9003cdf9.gif)

---
- In VSCode, clicking on a path segment reveals the corresponding segment (and *selected*) and its siblings at the top level vs. in this PR, we see the path segment's parent at the top level (and the segment is *unselected*):

**VSCode**
![image](https://user-images.githubusercontent.com/62660714/129807926-d4f95db7-9adc-493f-8ac6-45ccde6649f7.png)

**This PR**
![image](https://user-images.githubusercontent.com/62660714/129808008-73746223-79f7-4c4f-b89f-7a2daa030637.png)

---
The treeview seems to have a minimum height that looks a bit odd when there are not enough elements to populate the view. In VSCode the view seems to only take up the minimum height it needs (and changes height when nodes are expanded -- up to a fixed maximum height):

![image](https://user-images.githubusercontent.com/62660714/129808495-9536c4e9-489a-4fcf-afb3-2b9aea4122bf.png)

---
Very minor: the breadcrumbs is a 24px height but the viewContainer toolbar is 22px. In VSCode they are both 22px (I only noticed it via the broken toggle mentioned above 🙂) 

![image](https://user-images.githubusercontent.com/62660714/129808825-a23ab18a-e2e7-4350-9b41-e3d82168b2c3.png)

--- 
In VSCode, if I click on a path segment in the breadcrumbs and then click on a file inside that crumb, a new editor opens with that file in *preview* mode. In this PR, the file opens in normal editor mode.

**This PR**
![new-editor-open](https://user-images.githubusercontent.com/62660714/129809911-0ef7f159-eb1d-4175-82d7-608f49ee0d0c.gif)

**VSCode**
![vscode-open](https://user-images.githubusercontent.com/62660714/129810027-8e59343e-7763-436a-ae15-178a767636e7.gif)

colin-grant-work commented:
@kenneth-marut-work, thanks for taking a look. I believe I've addressed your comments.
 - the odd toggling was a consequence of not hooking into Phosphor's fit logic; I've added code that asks the tabbar's parent to fit everything to size.
 - added code that selects the node corresponding to the crumb clicked on.
 - changed the trees to non-virtualized so that they can have a definite height smaller than 200px if there's less content to render.
 - changed the height of the breadcrumbs area to 22px. You'll have to check in your fancy theme whether everything lines up now :-)
 - changed the opener to use `{preview: true}`

msujew reviewed (commented):
I managed to notice some weird behavior as well. When opening the file tree via the breadcrumbs and opening a folder in there, clicking outside of the popup does not close the popup. This leads to things like these:

![grafik](https://user-images.githubusercontent.com/4377073/130243250-3f99bc25-5f77-4c0e-8b6a-7200062a3cd7.png)

colin-grant-work commented:
> I managed to notice some weird behavior as well. When opening the file tree via the breadcrumbs and opening a folder in there, clicking outside of the popup does not close the popup.

@msujew, thanks for pointing this out. It pointed the way to some overly elaborate focus handling logic that conflicted with the expectations of the widgets being rendered in the popups. I've simplified that and used `focusout` events rather than `blur` to make things a little more flexible.

kenneth-marut-work commented on packages/core/src/browser/widgets/react-renderer.tsx:
```diff
@@ -19,6 +19,9 @@ import * as React from 'react';
 import * as ReactDOM from 'react-dom';
 import { Disposable } from '../../common';
 
+export type RendererHost = HTMLElement;
```
Looks like these declarations might make more sense in the `breadcrumb-popup-container` file?

kenneth-marut-work commented on packages/outline-view/src/browser/outline-view-widget.tsx:
```diff
@@ -27,9 +27,11 @@ import {
+import { Emitter, Mutable, UriSelection } from '@theia/core';
 import { CompositeTreeNode } from '@theia/core/lib/browser';
 import * as React from '@theia/core/shared/react';
+import { Range } from '@theia/editor/lib/browser/editor';
```
I think this is the only import from `editor` in this package if I'm not mistaken? Any way you can decouple editor from this package and get this type another way so you don't have to add the additional dependency?

kenneth-marut-work reviewed (commented):
A couple minor comments, also as we discussed offline the WorkspaceBreadcrumbsContribution doesn't seem to be doing what it should (prefixing the breadcrumbs with WS name in multiroot). But otherwise the functionality is very impressive and is working great for me, and thanks for addressing my previous comments as well 👍

colin-grant-work commented on packages/core/src/browser/breadcrumbs/breadcrumbs-renderer.tsx:
```diff
@@ -0,0 +1,233 @@
+import { CorePreferences } from '../core-preferences';
+import { Coordinate } from '../context-menu-renderer';
+
+interface Cancelable {
```
This needs to be an object so that there can be a shared reference to it on the class and in the method. If it were just a boolean, setting the member to `false` wouldn't affect the variable referred to be the previous method call that needs to be cancelled.

msujew reviewed (approved):
Looks good to me, great work 👍 

One minor nitpick: vscode uses a larger padding on the left side, which I think looks a bit nicer:

vscode:

![image](https://user-images.githubusercontent.com/4377073/131341588-1284c71e-8972-457f-8598-67b85633627f.png)

Theia:

![image](https://user-images.githubusercontent.com/4377073/131341621-e2cd620f-0d86-44d5-949a-07c705929f94.png)

vince-fugnitto commented on packages/core/src/browser/breadcrumbs/breadcrumbs-constants.ts:
```diff
@@ -0,0 +1,79 @@
+    /** A CSS class for the icon. */
+    readonly iconClass?: string;
+
+    /** CSS classes for the container */
```
```suggestion
    /** CSS classes for the container. */
```

vince-fugnitto commented on packages/core/src/browser/breadcrumbs/breadcrumbs-service.ts:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    /**
+     * Subscribe to this event emitter to be notifed when the breadcrumbs have changed.
```
minor typo:

```suggestion
     * Subscribe to this event emitter to be notified when the breadcrumbs have changed.
```

vince-fugnitto commented on packages/core/src/browser/breadcrumbs/breadcrumbs-service.ts:
```diff
@@ -0,0 +1,108 @@
+     * The URI is the URI of the editor the breadcrumbs have changed for.
+     */
+    get onDidChangeBreadcrumbs(): Event<URI> {
+        // This lazy subscription is to address problems in iversify's instantiation routine
```
minor typo:

```suggestion
        // This lazy subscription is to address problems in inversify's instantiation routine
```

vince-fugnitto commented on packages/filesystem/src/browser/breadcrumbs/filepath-breadcrumbs-contribution.ts:
```diff
@@ -0,0 +1,129 @@
+
+export const FilepathBreadcrumbType = Symbol('FilepathBreadcrumb');
+
+export interface FilepathBreadccrumbClassNameFactory {
```
typo:

```suggestion
export interface FilepathBreadcrumbClassNameFactory {
```

vince-fugnitto commented on packages/monaco/src/browser/monaco-outline-contribution.ts:
```diff
@@ -42,20 +41,17 @@ export class MonacoOutlineContribution implements FrontendApplicationContributio
-        });
+
+        // updateOutline and handleCurrentEditorChanged need to be called even when the outline view widget is closed
+        // in order to udpate breadcrumbs.
```
minor typo:

```suggestion
        // in order to update breadcrumbs.
```

vince-fugnitto commented on packages/outline-view/src/browser/outline-view-service.ts:
```diff
@@ -61,8 +61,10 @@ export class OutlineViewService implements WidgetFactory {
-            this.onDidChangeOutlineEmitter.fire(roots);
         }
+        // onDidChangeOutline needs to be fired even when the outline view widget is closed
+        // in order to udpate breadcrumbs.
```
minor typo:

```suggestion
        // in order to update breadcrumbs.
```

vince-fugnitto commented on packages/core/src/browser/breadcrumbs/breadcrumbs-service.ts:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    /**
+     * Subscribes to the onDidChangeBreadcrumbs events for all contributions
```
```suggestion
     * Subscribes to the onDidChangeBreadcrumbs events for all contributions.
```

vince-fugnitto commented:
I noticed the `perfect-scrollbar` styling might be off for the breadcrumbs compared to the rest of the application (and vscode).
I'm not sure if it is something we want handled initially by the pull-request but I was able to reproduce on `electron`:


https://user-images.githubusercontent.com/40359487/131505872-403d7af1-6ed5-4844-867e-dec2fcf8e309.mp4

colin-grant-work commented:
> I noticed the `perfect-scrollbar` styling might be off for the breadcrumbs compared to the rest of the application (and vscode).
> I'm not sure if it is something we want handled initially by the pull-request but I was able to reproduce on `electron`:

@vince-fugnitto The easiest (and, I think, most correct) way to solve that would be to remove the scroll bar from the breadcrumb popup container and require breacrumb providers to provide their own scroll if necessary - both of the current ones show tree widgets, so they come with their own scrollbars anyway. That also avoids duplication of scroll bars in case the rendered UI happens to include a scroll bar, as all widgets would. Does that sound reasonable?

vince-fugnitto commented:
> @vince-fugnitto The easiest (and, I think, most correct) way to solve that would be to remove the scroll bar from the breadcrumb popup container and require breacrumb providers to provide their own scroll if necessary - both of the current ones show tree widgets, so they come with their own scrollbars anyway. That also avoids duplication of scroll bars in case the rendered UI happens to include a scroll bar, as all widgets would. Does that sound reasonable?

I think that sounds reasonable yes :)

colin-grant-work commented:
> @vince-fugnitto The easiest (and, I think, most correct) way to solve that would be to remove the scroll bar from the breadcrumb popup container and require breacrumb providers to provide their own scroll if necessary - both of the current ones show tree widgets, so they come with their own scrollbars anyway. That also avoids duplication of scroll bars in case the rendered UI happens to include a scroll bar, as all widgets would. Does that sound reasonable?

Turns out I couldn't do that. I'd hoped the good styles were attached to widgets, but in fact they're attached to the shell, and dialogues, and now also breadcrumb popups. I think we could do that better, since I think we really just mean 'everywhere in the application,' but someone decided that that should be classed down to the shell. If you agree, @vince-fugnitto, we can probably simplify the perfect scrollbar CSS section by eliminating all of the #-referencing prefixes, but that might have unintended consequences (maybe the monaco overlays use PS and want different styles, e.g.? Perhaps best to put that off for another PR.

colin-grant-work commented:
If CI passes and no one objects, I'll merge this Wednesday or Thursday.
